[#149806558] Compose metric scraping and reporting to Datadog

46bit opened the pull request:
## What

We have implemented a scraper for the metrics of our Compose cluster (as used for production.) This uses the Scrapey library for Python to perform the HTTPS requests, extract the relevant elements and extract the RAM and Disk metrics for each host.

An app to be deployed on the PaaS runs the scraper and metric reporting periodically. Metrics are logged to Datadog:

* Metrics of the `in_use` (percentage used), `used` (GB used) and `total` (GB total) for RAM and Disk for every Compose host.
* Service check that the service is alive, and any failed scraping attempts.

For more information see the commit messages and the new `README.md`.

## How to review

* Probably best to review in conjunction with [paas-cf PR1036](https://github.com/alphagov/paas-cf/pull/1036) that deploys this app and create Datadog monitors to alert on this data. 

## Who can review

Not @46bit @timmow @henrytk @combor @chrisfarms

Commit:
Compose scraper which reports to Datadog

We have implemented a scraper for the metrics of our Compose cluster
(as used for production.) This uses the Scrapey library for Python to
perform the HTTPS requests, extract the relevant elements and extract
the RAM and Disk metrics for each host.

This is wrapped into an app to be deployed on the PaaS that performs
the scraping process periodically. Metrics are logged to Datadog.

The underlying HTML representation is non-semantic and seems likely to
change at a later date. We have broken down the scraping process into
`ClusterScraper` and `HostScraper` classes, which are designed to
resiliently extract as much information as possible even if one host
or one host's RAM information can't be extracted.

If information cannot be extracted, the scraper reports this to a
Datadog service check. Any missing data will be logged as a Warning.
If all data is missing (e.g., the page can no longer be scraped or no
hosts were extracted) this is logged as an Error.

A separate pull request on paas-cf integrates this into the PaaS:
https://github.com/alphagov/paas-cf/pull/1031

Commit:
README describing how to work with the scraper

dcarley committed:
Add TravisCI config

So that we don't *have* to run the tests locally when reviewing PRs. Using
Python 3.6 as specified in `runtime.txt` used by Cloud Foundry.

dcarley commented on cluster_spider.py:
```diff
@@ -0,0 +1,120 @@
+            callback=self.after_login)
+
+    def after_login(self, response):
+        url = 'https://app.compose.io/gds/clusters/5941cf9f859d2c0015000021'
```
Can we pass the organisation name and cluster ID in as environment variables so that they're not hardcoded?

Partly for re-usability and partly because I'm not sure we should advertise our cluster ID - I don't think it's super sensitive, certainly not enough to scrub history, but we should avoid baking it in.

chrisfarms replied:
👍

dcarley commented on utils.py:
```diff
@@ -0,0 +1,83 @@
+    """
+    tags = [
+        'cluster:{}'.format(cluster_name),
+        'deployment:{}'.format(os.environ['DEPLOY_ENV'])
```
We should document in the README that this is required. It might be good to call it something more agnostic of our setup, like `DEPLOYMENT_TAG`.

chrisfarms replied:
Don't feel strongly about renaming it tbh... but I think I'll change it to just allow setting a csv of tags from environment to leave it up to the user.

`DATADOG_TAGS="deployment:prod,whatever:treva"`

dcarley commented on main.py:
```diff
@@ -0,0 +1,12 @@
+    utils.init_datadog()
+    utils.Report.ok()
+    subprocess.Popen(
+        ["scrapy", "runspider", "cluster_spider.py"], env=os.environ)
```
I think we should check for the environment variables are startup and fail with an error if we don't have everything that we need. The error that you get at the moment is quite confusing:
```
➜  paas-compose-scraper git:(149806558-compose-metric-alerting) ✗ DEPLOY_ENV=foo pipenv run python3 main.py
OK
Traceback (most recent call last):
  File "main.py", line 9, in <module>
    utils.Report.ok()
  File "/Users/dcarley/projects/paas/paas-compose-scraper/utils.py", line 79, in ok
    report_health(CheckStatus.OK, message)
  File "/Users/dcarley/projects/paas/paas-compose-scraper/utils.py", line 69, in report_health
    check=check, host_name=host, status=status, message=message, tags=tags)
  File "/Users/dcarley/.local/share/virtualenvs/paas-compose-scraper-bR8x88GK/lib/python3.6/site-packages/datadog/api/service_checks.py", line 39, in check
    return super(ServiceCheck, cls)._trigger_action('POST', 'check_run', **params)
  File "/Users/dcarley/.local/share/virtualenvs/paas-compose-scraper-bR8x88GK/lib/python3.6/site-packages/datadog/api/resources.py", line 215, in _trigger_action
    return APIClient.submit(method, name, params)
  File "/Users/dcarley/.local/share/virtualenvs/paas-compose-scraper-bR8x88GK/lib/python3.6/site-packages/datadog/api/api_client.py", line 87, in submit
    raise ApiNotInitialized("API key is not set."
datadog.api.exceptions.ApiNotInitialized: API key is not set. Please run 'initialize' method first.
```


Can we only pass over the environment variables that we need, or is a full environment needed?

chrisfarms replied:
The passing of the entire environment is mainly just an artifact of the `subprocess` method of executing the scraper.

However the shouty stacktrace errors have been addressed by other changes. (an env helper to gracefully bow out on missing required vars and making setting tags optional via `DATADOG_TAGS` envvar

chrisfarms replied:
Heh, ok think I misunderstood ... willfix 👍

dcarley replied:
I was thinking about whether we need to pass over all of `os.environ` because I was finding it difficult to understand, at a high level without reading further down the code, what environment variables were required to run. It felt like they could be function arguments instead. It sounds like the env helper will help though.

dcarley commented on responses/fixtures/compose.html:
```diff
@@ -0,0 +1,2 @@
+<!DOCTYPE html><html lang="en"><head><meta charset="utf-8" /><meta content="IE=edge" http-equiv="X-UA-Compatible" /><meta name="csrf-param" content="authenticity_token" />
+<meta name="csrf-token" content="GC78xs7FWdBjRzp5Ds9QIHkWKHK+78cjkktIj27k2FCLibWTd2rWG7scHAYVNIadkZMZv51JyavfOYUQIr0NxA==" /><meta content="True" name="HandheldFriendly" /><meta content="width=device-width, initial-scale=1.0" name="viewport" /><title>Compose</title><link rel="shortcut icon" type="image/x-icon" href="https://d11f55tj5eo9e5.cloudfront.net/assets/megatron/favicon.ico" /><link rel="m ...
```
Similar to a previous comment about variablising the org name and cluster ID, can we redact them from the fixture file?

chrisfarms replied:
Yep 🙌  ...did this when adding `COMPOSE_ACCOUNT_NAME` and `COMPOSE_CLUSTER_ID` env vars

dcarley commented on main.py:
```diff
@@ -0,0 +1,12 @@
+while True:
+    utils.init_datadog()
+    utils.Report.ok()
+    subprocess.Popen(
```
Opening a new subprocess feels a bit odd. I'm kind of surprised that the initialised DataDog library is available from the subprocess. Is it possible to use scrapy as a library from the same Python process?

46bit replied:
Agreed, and the Datadog library is reinitialised inside the subprocess.

chrisfarms replied:
Seem like it uses twisted under the hood ... so might not be simplest library import without the helper runner, but I'll have a poke around.

I think another issue with this loop is that Popen is async ... so as it stands it will run a new process every 5mins regardless of whether the last one complete ... which could be leak (although unlikely given the 5min timer).

That first `init_datadog` call is just to allow for the `Report.ok()` to phone in I believe ... not the main reporting.

46bit replied:
I think it would be best if `utils.Report.ok()` was only called if we successfully got all the data. But at present there's no easy way to code that.

dcarley commented on settings.py:
```diff
@@ -0,0 +1,7 @@
+SPIDER_MODULES = ['compose.spiders']
+NEWSPIDER_MODULE = 'compose.spiders'
+ROBOTSTXT_OBEY = True
+CHECK_INTERVAL_SECONDS = 600
```
Thanks for making configurable 👍

46bit replied:
Candidate for optional environment variable?

chrisfarms replied:
Yeah, why not... I'll make it configurable from environment

dcarley reviewed (changes requested):
Firstly well done 👏 

I tested this with #1031 - I know you said not to until this PR was merged, but it was easier to run that way. Few things that didn't fit into line-by-line review:

1. I've pushed a commit to run the tests automatically in Travis and have configured the relevant hooks. Can someone check that they're happy with it?
1. I've noticed that one host is being reported twice for each 10min interval. Anyone know why?

    ![screen shot 2017-09-14 at 15 18 42](https://user-images.githubusercontent.com/260438/30434814-1830172c-9960-11e7-8ebe-39091b423429.png)
1. We need to create a PR for https://github.com/alphagov/paas-credentials/compare/compose-billing
1. `requests` is logging in DEBUG mode, as seen from `cf logs paas-compose-scraper`, but I can't see it being explicitly configured. Do we need it? Should we disable it to make the logs easier to read and retain more history about the scraping?

dcarley commented:
NB: I used `SELF_UPDATE_PIPELINE=false` when testing. I didn't re-run the paas-bootstrap pipelines with `ENABLE_DATADOG=true` because I didn't need the host metrics to test this functionality.

46bit commented:
I don't think I should review this, but I discussed self-updating pipelines with @alext earlier. `SELF_UPDATE_PIPELINE=false` is needed because `ENABLE_DATADOG` is not a [flag forwarded through](https://github.com/alphagov/paas-cf/blob/master/concourse/pipelines/create-cloudfoundry.yml#L341-L348) `self-update-pipelines` when it re-makes the pipeline config.

If we forwarded `ENABLE_DATADOG` then I believe a dev Concourse would keep Datadog enable until you explicitly. There's a few ways one can work with Concourse, from custom Makefiles to providing flags like we do here, and it's tricky to keep them all available. I think @alext is going to think a bit about this :-)

Commit:
Use environment to pass account name and cluster

We shouldn't really advertise our cluster id and it would be better if
it was configurable.

This pulls the values from COMPOSE_ACCOUNT_NAME and COMPOSE_CLUSTER_ID
environment variables instead.

Commit:
Allow setting multiple tags from environment

Rather than forcing anyone using this library to use our
`deployment:$DEPLOY_ENV` convention, just allow setting tags from a list
using DATADOG_TAGS environment variable

Commit:
Check environment variables on startup

Previously the required environment variables were only fetched as
required by the subprocess. This gave poor error messages when they were
missing.

This changes it so that required variables are checked during execution
of the parent so we can bail with better errors early.

We also expose the CHECK_INTERVAL_SECONDS environment variable to make
it simpler to alter the time between scraping attempts

chrisfarms commented:
Isn't the "ok" check somewhat redundant? ... The monitor could just alert
on "no data" for X periods to the same effect.... Unless I misunderstood
it's purpose?

On 15 Sep 2017 15:57, "Michael Mokrysz" <notifications@github.com> wrote:

> *@46bit* commented on this pull request.
> ------------------------------
>
> In main.py
> <https://github.com/alphagov/paas-compose-scraper/pull/1#discussion_r139171424>
> :
>
> > @@ -0,0 +1,12 @@
> +import settings
> +import os
> +import subprocess
> +import utils
> +from time import sleep
> +
> +while True:
> +    utils.init_datadog()
> +    utils.Report.ok()
> +    subprocess.Popen(
>
> I think it would be best if utils.Report.ok() was only called if we
> successfully got all the data. But at present there's no easy way to code
> that.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/alphagov/paas-compose-scraper/pull/1#discussion_r139171424>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACzYXV6-845rjudnBR4ywgwLYpUT7AGks5sipA0gaJpZM4PTAuq>
> .
>

Commit:
Run as a script using CrawlRunner rather than subprocess

Shelling out to a scrapey subprocess was giving poor errors and made
exiting the subprocess via SIGINT problimatic.

This changes to use the CrawlRunner helper to run the crawler in the
twisted reactor event loop manually.

As part of this change the global settings config was retired in favour
of building the settings dict within main to make a clear single
entrypoint for mapping of environment variables to settings.

The LOG_LEVEL setting was also exposed to environment and now defaults
to 'WARNING' to reduce the noise.

dcarley commented:
Changes look good! Thanks @46bit and @chrisfarms !
